cupy/core/_fusion_analysis.py

```diff
@@ -0,0 +1,591 @@
+
+        # Parse Inputs.
+        nin = ufunc.nin
+        nout = ufunc.nout
+        dtype = kwargs.pop('dtype', None)
+
+        if 'out' in kwargs and len(args) > nin:
+            raise ValueError(
+                'cannot specify \'out\' as both a positional and '
+                'keyword argument')
+
+        in_params = [self._from_arraylike_interface(x) for x in args[:nin]]
+        out_params = [
+            self._from_interface(x)
+            for x in args[nin:] + (kwargs.pop('out', None),)
+            if x is not None
+        ]
+        params = in_params + out_params
+
+        if len(kwargs) > 0:
+            raise TypeError('Wrong arguments {}'.format(kwargs))
+        if len(in_params) != nin or len(out_params) > nout:
+            raise ValueError('Invalid number of arguments')
+        if not all([isinstance(v, _FusionCudaArray) for v in out_params]):
+            raise TypeError('Return arrays must be of ArrayType')
+
+        # Check for inplace operation.
+        for i, out_param1 in enumerate(out_params):
+            for out_param2 in out_params[:i]:
+                if out_param1.memory == out_param2.memory:
+                    # NumPy does not raise this error.
+                    raise ValueError('Outputs of ufunc must not share memory')
+
+        for i, in_param in enumerate(in_params):
+            should_copy = any([
+                in_param.memory == out_param.memory and in_param != out_param
+                for out_param in out_params
+            ])
+            if should_copy:
+                in_params[i] = self._from_arraylike_interface(
+                    self.call_ufunc(
+                        core.elementwise_copy,
+                        self._to_interface(in_param)))
+
+        # Broadcast shapes
+        out_rshape = _broadcast_shapes([p.rshape for p in params])
+        out_ashape = [None for _ in range(len(out_rshape))]
+
+        for p in params:
+            for axis in range(-p.ndim, 0):
+                if p.rshape[axis] == out_rshape[axis]:
+                    out_ashape[axis] = p.ashape[axis]
+
+        assert all([dim is not None for dim in out_ashape])
+        out_ashape = tuple(out_ashape)
+
+        # Broadcast input params and make their views.
+        for i, p in enumerate(in_params):
+            for axis in range(-p.ndim, 0):
+                if p.rshape[axis] == out_rshape[axis]:
+                    self.shape_constraints.add_eq_constraint(
+                        p.ashape[axis], out_ashape[axis])
+                elif p.rshape[axis] == 1:
+                    self.shape_constraints.add_const_constraint(
+                        p.ashape[axis], 1)
+                else:
+                    assert False
+            if isinstance(p, _FusionCudaArray) and p.rshape != out_rshape:
+                # Broadcst input if needed.
+                in_params[i] = self.vc.broadcast_to(p, out_ashape, out_rshape)
+
+        # Get operation code from dtypes.
```

Comment:
```suggestion
        # Get operation code from output dtype.
```

Reply:
If `dtype` is `None`, the operation code is determined from the dtypes of input/output arrays.